### PR TITLE
Force controllers to use state in preference to specs.

### PR DESCRIPTION
**Pull request by brandonbloom**
Exceptions are initialize & refresh.

Further attempts to mitigate #270.

### CODE REVIEW

**Comment by brandonbloom** [comment]
@BenElgar I need to do a fair bit of testing on this, but it should be reviewable.

**Comment by brandonbloom** [comment]
OK, did a bunch of testing & this seems good. Going to merge. Post-hoc review welcome.

File: internal/providers/docker/components/container/build.go [hunk 1]
```diff
@@ -22,31 +25,37 @@ import (
 )
 
 func (c *Container) Build(ctx context.Context, input *api.BuildInput) (*api.BuildOutput, error) {
-	if c.canBuild() {
-		if err := c.buildImage(ctx); err != nil {
+	if c.State.Image.Spec == "" {
+		// SEE NOTE: [MIGRATE_CONTAINER_STATE].
+		return nil, errors.New("refresh needed")
```
Comment (BenElgar) [inline internal/providers/docker/components/container/build.go:30]: When will this refresh happen? Do people need to run the manual `exo refresh` command?

**Review by BenElgar — COMMENTED** [review]
It all LGTM but I'm not quite sure I understand the point. Isn't the issue that a new spec doesn't appropriately replace an invalid state? How does using the state in preference to the spec help with that?

**Comment by brandonbloom** [comment]
> Isn't the issue that a new spec doesn't appropriately replace an invalid state? How does using the state in preference to the spec help with that?

Prior to this change, every operation – including removing a component – relied on being able to parse the spec. Since specs are user-provided, they are much more likely to have syntax errors and other problems. State being incorrect, on the other hand, is a sign of a bug in the provider/component.

**Pull request merged**